### ISSUE TITLE
FocusScope causing unintended focus trap

### ISSUE THREAD

**Issue by Andarist** [adobe/react-spectrum#1570]
# 🐛 Bug Report 

There is a way to end up with the focus being trapped due to the logic in `useRestoreFocus`.

1. Have a FocusScope with autoFocus that records an unmounting element as `nodeToRehydrate`
2. Try to tab away from that FocusScope to the next element
3. `useRestoreFocus` does this:
    - [checks if the `focusedElement` is in the current scope](https://github.com/adobe/react-spectrum/blob/739d7f1eb4ea2dbb00190c31fc709164c1ca6f35/packages/%40react-aria/focus/src/FocusScope.tsx#L340) - it is so the logic continues
    - [checks if the nextElement is in the scope and if we have nodeToRehydrate set](https://github.com/adobe/react-spectrum/blob/739d7f1eb4ea2dbb00190c31fc709164c1ca6f35/packages/%40react-aria/focus/src/FocusScope.tsx#L353) - `nextElement` is not in the scope and we have `nodeToRehydrate` set
    - [it searches for the next element from the **detached** `nodeToRehydrate` ](https://github.com/adobe/react-spectrum/blob/739d7f1eb4ea2dbb00190c31fc709164c1ca6f35/packages/%40react-aria/focus/src/FocusScope.tsx#L357-L359) and finds a yet another detached element
    - [it tries to focus a detached element](https://github.com/adobe/react-spectrum/blob/739d7f1eb4ea2dbb00190c31fc709164c1ca6f35/packages/%40react-aria/focus/src/FocusScope.tsx#L363-L364) [while preventing the default behavior](https://github.com/adobe/react-spectrum/blob/739d7f1eb4ea2dbb00190c31fc709164c1ca6f35/packages/%40react-aria/focus/src/FocusScope.tsx#L361)
    - focusing detached element fails silently and the focus is left in place, this effectively prevents users from tabbing through a page

**Note:** In our situation real-life we still want to use `restoreFocus` as we need to get back to the "trigger" element but at the same time it should be possible to tab away from the element using the default browser behavior. In our situation, our app is contained within an iframe and the user is expected to use the tab key to just escape the iframe and get back to the host page.

## 🤔 Expected Behavior

Being able to tab away from the element.

## 😯 Current Behavior

Can't tab away from a FocusScope 

## 💻 Code Sample

https://codesandbox.io/s/agitated-lovelace-cqrt4?file=/src/App.js

**Comment by ktabors** [comment]
I was not able to reproduce this in chrome on osx, but I was with Safari and Firefox. My reproduce I mean that in Firefox and Safari the focus is contained and does not escape like it should.

However, when I inspect the code I see the same code execution path in Chrome and Firefox, which is no nextElement is found and the focusedElement.blur() is called. https://github.com/adobe/react-spectrum/blob/739d7f1eb4ea2dbb00190c31fc709164c1ca6f35/packages/%40react-aria/focus/src/FocusScope.tsx#L367 which is different than you report seeing.

**Comment by Andarist** [comment]
I was definitely able to reproduce this in Chrome on OSX. I can try to record what I'm experiencing tomorrow if that's needed~. There is a chance that what you are observing is a separate issue - have you been following my repro steps closely?

Note that I'm willing to work on a fix but a proper fix (for my exact scenario) is probably more involved than what I've temporarily applied in our project:
```diff
diff --git a/node_modules/@react-aria/focus/dist/module.js b/node_modules/@react-aria/focus/dist/module.js
index 9bdb9ef..8798984 100644
--- a/node_modules/@react-aria/focus/dist/module.js
+++ b/node_modules/@react-aria/focus/dist/module.js
@@ -337,6 +337,11 @@ function $c9e8f80f5bb1841844f54e4ad30b$var$useRestoreFocus(scopeRef, restoreFocu
       let nextElement = e.shiftKey ? walker.previousNode() : walker.nextNode(); // If there is no next element, or it is outside the current scope, move focus to the
       // next element after the node to restore to instead.

+      if (!document.contains(nodeToRestore)) {
+        nodeToRestore = null
+      }
+
       if ((!nextElement || !$c9e8f80f5bb1841844f54e4ad30b$var$isElementInScope(nextElement, scope)) && nodeToRestore) {
         walker.currentNode = nodeToRestore; // Skip over elements within the scope, in case the scope immediately follows the node to restore.

```

So that's why I've raised this issue first - to discuss possible approaches etc.

**Comment by majornista** [comment]
I think I may have a fix for this in https://github.com/adobe/react-spectrum/pull/1493: https://github.com/adobe/react-spectrum/pull/1493/commits/cc9656850ff1af0ca58b79932b2839fed67797df.

**Comment by majornista** [comment]
You can reproduce this problem with the `useFocusManager` example on the documentation site: https://react-spectrum.adobe.com/react-aria/FocusScope.html#usefocusmanager-example. Keyboard focus gets trapped on the "Paste" button when tabbing through the toolbar items, and focus blurs to the document.body when Shift+Tabbing away from the "Cut" button. The blur to the document.body behaves like an empty tab stop, with no item showing focus.

**Comment by Andarist** [comment]
I'm pretty sure that this won't fix my problem. In the steps that I've listed in 3c the `nodeToRehydrate` is there but it is a detached node (so it's different than `document.body`).

**Comment by Andarist** [comment]
I've recorded the repro steps for this bug with some commentary about the code: https://youtu.be/f3LCtetuaQc . Note that I'm being able to reproduce this in any browser (Chrome/FireFox/Safari) and this doesn't seem to be browser-specific.

**Comment by probablyup** [comment]
@Andarist I'm seeing this behavior in Storybook as well, interestingly. If a story has a `FocusScope` in it with `contain`, it somehow leaks out of the story iframe and interrupts focus management in the rest of the storybook UI.

**Comment by Andarist** [comment]
Yeah, cross-frame focus management is really tricky because there are no means to actually track this stuff from a single "realm" (aka frame, in this context). So I'm somewhat not surprised that you are running into issues like this. I was using React Aria in an isolated iframe and we had to hack around and sync some information back and forth using `postMessage` to manage some stuff properly.

**Comment by probablyup** [comment]
This property might be useful: https://developer.mozilla.org/en-US/docs/Web/API/UIEvent/view

If event.view !== window, ignore it
